File: online-judges-and-contests/codewars/2kyu-power-tower/power-tower.go
package kata

import (
	"math/big"
	"sync"
)

var (
	phiCache = make(map[uint64]uint64)
	mu       sync.RWMutex
)

func Tower(base, height uint64, modulus uint32) uint32 {
	switch {
	case modulus == 1:
		return 0
	case base == 1 || height == 0:
		return 1
	case height == 1:
		return uint32(base % uint64(modulus))
	}

	baseBig := big.NewInt(int64(base))
	return uint32(calculateGrowth(baseBig, height, uint64(modulus)).Uint64() % uint64(modulus))
}

func calculateGrowth(base *big.Int, height uint64, modulus uint64) *big.Int {
	if height == 1 || modulus == 1 {
		return calculateOperation(base, big.NewInt(int64(modulus)))
	}

	nextModulus := totientEuler(modulus)
	nextGrowth := calculateGrowth(base, height-1, nextModulus)
	return towerModPow(base, nextGrowth, big.NewInt(int64(modulus)))
}

func calculateOperation(input, divisor *big.Int) *big.Int {
	fourTimesDivisor := new(big.Int).Mul(divisor, big.NewInt(4))
	if input.Cmp(fourTimesDivisor) < 0 {
		return input
	}
	remainder := new(big.Int).Mod(input, divisor)
	threeTimesDivisor := new(big.Int).Mul(divisor, big.NewInt(3))
	return new(big.Int).Add(remainder, threeTimesDivisor)
}

func towerModPow(base, exp, mod *big.Int) *big.Int {
	result := big.NewInt(1)
	zero := big.NewInt(0)
	one := big.NewInt(1)

	for exp.Cmp(zero) != 0 {
		if new(big.Int).And(exp, one).Cmp(one) == 0 {
			result = calculateOperation(new(big.Int).Mul(result, base), mod)
		}
		base = calculateOperation(new(big.Int).Mul(base, base), mod)
		exp = new(big.Int).Rsh(exp, 1)
	}
	return result
}

func totientEuler(n uint64) uint64 {
	mu.RLock()
	if result, exists := phiCache[n]; exists {
		mu.RUnlock()
		return result
	}
	mu.RUnlock()

	originalN := n
	result := n
	for i := uint64(2); i*i <= n; i++ {
		if n%i == 0 {
			for n%i == 0 {
				n /= i
			}
			result -= result / i
		}
	}
	if n > 1 {
		result -= result / n
	}

	mu.Lock()
	phiCache[originalN] = result
	mu.Unlock()

	return result
}

File: online-judges-and-contests/codewars/2kyu-power-tower/power-tower.py
import threading

phi_cache = {}
cache_lock = threading.Lock()

def euler_totient(n):
    with cache_lock:
        if n in phi_cache:
            return phi_cache[n]
    
    original_n = n
    result = n
    i = 2
    while i * i <= n:
        if n % i == 0:
            while n % i == 0:
                n //= i
            result -= result // i
        i += 1
    if n > 1:
        result -= result // n
    
    with cache_lock:
        phi_cache[original_n] = result
    
    return result

def calculate_operation(input, divisor):
    four_times_divisor = divisor * 4
    if input < four_times_divisor:
        return input
    remainder = input % divisor
    three_times_divisor = divisor * 3
    return remainder + three_times_divisor

def tower_mod_pow(base, exp, mod):
    result = 1
    while exp != 0:
        if exp % 2 == 1:
            result = calculate_operation(result * base, mod)
        base = calculate_operation(base * base, mod)
        exp //= 2
    return result

def calculate_growth(base, height, modulus):
    if height == 1 or modulus == 1:
        return calculate_operation(base, modulus)
    
    next_modulus = euler_totient(modulus)
    next_growth = calculate_growth(base, height - 1, next_modulus)
    return tower_mod_pow(base, next_growth, modulus)

def tower(base, height, modulus):
    if modulus == 1:
        return 0
    if base == 1 or height == 0:
        return 1
    if height == 1:
        return base % modulus
    
    return calculate_growth(base, height, modulus) % modulus
